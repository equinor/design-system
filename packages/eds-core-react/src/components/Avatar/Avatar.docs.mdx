import { Links } from './../../../.storybook/components'
import { Meta, Primary, Canvas, Controls } from '@storybook/blocks'
import * as ComponentStories from './Avatar.stories'

<Meta of={ComponentStories} />

# Avatar

Avatars use images to visually represent people or companies.<br />
The avatar does not exist in Figma as a separate component.

<Links 
    documentationUrl="https://eds.equinor.com/0b0c666ab/p/96f1b1-image-placeholder/b/85d6bc"
    sourceUrl="https://github.com/equinor/design-system/blob/develop/packages/eds-core-react/src/components/Avatar/Avatar.tsx"
    npmUrl="https://www.npmjs.com/package/@equinor/eds-core-react"
/>

<Primary />
<Controls />

## Usage

<ul>
    <li>Place avatars alongside additional information about the people or companies the avatars represent.</li>
    <li>Avatars can be displayed in lists that are searchable and selectable.</li>
    <li>Use inline avatars to mention people in a sentence.</li>
</ul>

```tsx
import { Avatar } from '@equinor/eds-core-react'

<Avatar />
```

## Examples

### Sizes

<Canvas of={ComponentStories.Size} />

### Disabled

<Canvas of={ComponentStories.Disabled} />

### Context

<Canvas of={ComponentStories.Context} />
