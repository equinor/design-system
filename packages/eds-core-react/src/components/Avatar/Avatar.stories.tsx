import { Story, ComponentMeta } from '@storybook/react'
import {
  Avatar,
  AvatarProps,
  Button,
  Card,
  Divider,
  Icon,
  Typography,
} from '../..'
import { Stack } from './../../../.storybook/components'
import page from './Avatar.docs.mdx'

export default {
  title: 'Data Display/Avatar',
  component: Avatar,
  parameters: {
    docs: {
      page,
    },
  },
} as ComponentMeta<typeof Avatar>

const iconSize = 24

export const Introduction: Story<AvatarProps> = (args) => (
  <Stack>
    <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} {...args} />
  </Stack>
)

export const Size: Story<AvatarProps> = () => (
  <Stack>
    <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} size={16} alt="avatar" />
    <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} size={24} alt="avatar" />
    <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} size={32} alt="avatar" />
    <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} size={40} alt="avatar" />
    <Avatar src={'https://i.imgur.com/UM3mrju.jpg'} size={48} alt="avatar" />
  </Stack>
)

export const Disabled: Story<AvatarProps> = () => (
  <Stack>
    <Avatar
      src={'https://i.imgur.com/UM3mrju.jpg'}
      disabled
      size={48}
      alt="avatar"
    />
  </Stack>
)

export const Context: Story<AvatarProps> = () => (
  <>
    <Stack style={{ background: '#ebebeb', padding: '32px' }}>
      <Card style={{ width: '50%' }}>
        <Card.Header>
          <Card.HeaderTitle>
            <Typography variant="h4">ALL</Typography>
          </Card.HeaderTitle>
        </Card.Header>
        <Card.Header>
          <Avatar
            alt="Kitten"
            src="https://i.imgur.com/UM3mrju.jpg"
            size={40}
          />
          <Card.HeaderTitle>
            <Typography variant="h5">Title goes here</Typography>
            <Typography variant="body_short">
              Ut enim ad minim veniam, quis nostrud exercitation.
            </Typography>
            <Divider style={{ width: '100%' }} variant="small" />
          </Card.HeaderTitle>
        </Card.Header>
        <Card.Header>
          <Avatar
            alt="Kitten"
            src="https://i.imgur.com/UM3mrju.jpg"
            size={40}
          />
          <Card.HeaderTitle>
            <Typography variant="h5">Title goes here</Typography>
            <Typography variant="body_short">
              Ut enim ad minim veniam, quis nostrud exercitation.
            </Typography>
            <Divider style={{ width: '100%' }} variant="small" />
          </Card.HeaderTitle>
        </Card.Header>
        <Card.Header>
          <Avatar
            alt="Kitten"
            src="https://i.imgur.com/UM3mrju.jpg"
            size={40}
          />
          <Card.HeaderTitle>
            <Typography variant="h5">Title goes here</Typography>
            <Typography variant="body_short">
              Ut enim ad minim veniam, quis nostrud exercitation.
            </Typography>
          </Card.HeaderTitle>
        </Card.Header>
      </Card>
    </Stack>

    <Stack style={{ background: '#ebebeb', padding: '32px' }}>
      <Card variant="info">
        <Card.Header>
          <Card.HeaderTitle>
            <Typography variant="h5">Title goes here</Typography>
            <Typography variant="body_short">Body short</Typography>
          </Card.HeaderTitle>
          <Avatar
            alt="Kitten"
            src="https://i.imgur.com/UM3mrju.jpg"
            size={40}
          />
        </Card.Header>
      </Card>
      <Card variant="warning">
        <Card.Header>
          <Avatar
            alt="Kitten"
            src="https://i.imgur.com/UM3mrju.jpg"
            size={40}
          />
          <Card.HeaderTitle>
            <Typography variant="h5">Title goes here</Typography>
            <Typography variant="caption">Caption</Typography>
          </Card.HeaderTitle>
        </Card.Header>
      </Card>
      <Card variant="danger">
        <Card.Header>
          <Avatar
            alt="Kitten"
            src="https://i.imgur.com/UM3mrju.jpg"
            size={40}
          />
          <Card.HeaderTitle>
            <Typography variant="h5">Title goes here</Typography>
            <Typography variant="caption">Caption</Typography>
          </Card.HeaderTitle>
          <Button variant="ghost_icon">
            <Icon
              name="more_vertical"
              title="more action"
              size={iconSize}
            ></Icon>
          </Button>
        </Card.Header>
      </Card>
    </Stack>
  </>
)
