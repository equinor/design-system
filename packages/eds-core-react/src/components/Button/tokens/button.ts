import { tokens } from '@equinor/eds-tokens'
import { ButtonToken } from '../Button.types'

const {
  typography: {
    navigation: { button: buttonTypography },
  },
  colors: {
    interactive: {
      disabled__text: { rgba: disabledTextColor },
      disabled__border: { rgba: disabledBorderColor },
      disabled__fill: { rgba: disabledColor },
    },
  },
  spacings: {
    comfortable: { medium },
  },
  shape: {
    button: { minHeight: buttonHeight, borderRadius: buttonBorderRadius },
    _modes: {
      compact: {
        button: { minHeight: compactButtonHeight },
      },
    },
  },
  interactions: { focused: outline },
  clickbounds: {
    default__base: clicboundHeight,
    compact__standard: compactClickboundHeight,
  },
} = tokens

export const button: ButtonToken = {
  background: 'transparent',
  height: `var(--eds_button__height, ${buttonHeight})`,
  typography: {
    ...buttonTypography,
    textAlign: 'center',
    fontSize: `var(--eds_button__font_size, ${buttonTypography.fontSize})`,
  },
  border: {
    type: 'border',
    width: `var(--eds_button__border_width, 1px)`,
    color: 'transparent',
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
    style: 'solid',
  },
  spacings: {
    top: 'var(--eds_button__padding_y, 0)',
    bottom: 'var(--eds_button__padding_y, 0)',
    left: `var(--eds_button__padding_x, ${medium})`,
    right: `var(--eds_button__padding_x, ${medium})`,
  },
  clickbound: {
    height: clicboundHeight,
    width: '100%',
    offset: {
      top: `${(parseInt(clicboundHeight) - parseInt(buttonHeight)) / 2 + 1}px`,
      left: '0',
    },
  },
  entities: {
    icon: {
      height: 'var(--eds_button__icon__size, 24px)',
      width: 'var(--eds_button__icon__size, 24px)',
    },
  },
  states: {
    hover: {
      border: {
        type: 'border',
        width: '1px',
        color: 'transparent',
        radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
        style: 'solid',
      },
    },
    focus: {
      outline: {
        type: 'outline',
        offset: '2px',
        style: 'dashed',
        color: outline.color,
        width: outline.width,
      },
    },
    disabled: {
      background: disabledColor,
      border: {
        type: 'border',
        width: '1px',
        color: disabledBorderColor,
        style: 'solid',
      },
      typography: {
        ...buttonTypography,
        color: disabledTextColor,
        textAlign: 'center',
      },
    },
  },
  modes: {
    compact: {
      height: `var(--eds_button__height_compact, ${compactButtonHeight})`,
      spacings: {
        top: 'var(--eds_button__padding_y_compact, 0)',
        bottom: 'var(--eds_button__padding_y_compact, 0)',
      },
      clickbound: {
        height: compactClickboundHeight,
        width: '100%',
        offset: {
          top: `${
            (parseInt(compactClickboundHeight) -
              parseInt(compactButtonHeight)) /
              2 +
            1
          }px`,
          left: '0',
        },
      },
    },
  },
}
