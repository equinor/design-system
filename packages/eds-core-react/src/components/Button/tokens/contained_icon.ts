import { tokens } from '@equinor/eds-tokens'
import { mergeDeepRight } from 'ramda'

import {
  primary as primaryContained,
  secondary as secondaryContained,
  danger as dangerContained,
} from './contained'
import { ButtonToken } from '../Button.types'

const {
  clickbounds: {
    default__base: clicboundHeight,
    compact__standard: compactClickboundHeight,
  },
  shape,
} = tokens

const contained_icon = {
  height: shape.icon_button.minHeight,
  width: shape.icon_button.minWidth,
  border: {
    width: '0px',
    radius: '50%',
  },
  spacings: { left: '0', right: '0' },
  clickbound: {
    width: clicboundHeight,
    offset: {
      top: `${
        (parseInt(clicboundHeight) - parseInt(shape.icon_button.minWidth)) / 2
      }px`,
      left: `${
        (parseInt(clicboundHeight) - parseInt(shape.icon_button.minWidth)) / 2
      }px`,
    },
  },
  states: {
    hover: {
      border: {
        width: '0px',
        radius: '50%',
      },
    },
  },
  modes: {
    compact: {
      height: shape._modes.compact.icon_button.minHeight,
      width: shape._modes.compact.icon_button.minWidth,
      clickbound: {
        width: compactClickboundHeight,
        offset: {
          top: '0',
          left: `${
            (parseInt(compactClickboundHeight) -
              parseInt(shape._modes.compact.icon_button.minWidth)) /
            2
          }px`,
        },
      },
    },
  },
}

export const primary = mergeDeepRight(
  primaryContained,
  contained_icon,
) as ButtonToken

export const secondary = mergeDeepRight(
  secondaryContained,
  contained_icon,
) as Partial<ButtonToken>
export const danger = mergeDeepRight(
  dangerContained,
  contained_icon,
) as Partial<ButtonToken>
