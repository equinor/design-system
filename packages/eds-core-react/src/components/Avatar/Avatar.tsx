import { forwardRef, HTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { bordersTemplate } from '@equinor/eds-utils'
import { avatar as tokens } from './Avatar.tokens'

type StyledAvatarProps = {
  $size: number
}

const StyledAvatar = styled.div<StyledAvatarProps>`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-shrink: 0;
  overflow: hidden;
  ${bordersTemplate(tokens.border)};
  ${({ $size }) => css`
    height: ${$size}px;
    width: ${$size}px;
  `}
`

type StyledImageProps = {
  alt: string
  src: string
  $disabled: boolean
}

const StyledImage = styled.img<StyledImageProps>`
  height: 100%;
  text-align: center;
  color: transparent;

  ${({ $disabled }) =>
    $disabled &&
    css`
      opacity: ${tokens.states.disabled.opacity};
    `};
`

export type AvatarProps = {
  alt: string
  /** Image source
   @default null */
  src?: string
  /** Avatar size
  @default 24 */
  size?: 16 | 24 | 32 | 40 | 48
  /** @default false */
  disabled?: boolean
} & HTMLAttributes<HTMLDivElement>

export const Avatar = forwardRef<HTMLDivElement, AvatarProps>(function Avatar(
  { src = null, alt, size = 24, disabled = false, ...rest },
  ref,
) {
  return (
    <StyledAvatar $size={size} ref={ref} {...rest}>
      <StyledImage src={src} alt={alt} $disabled={disabled} />
    </StyledAvatar>
  )
})

Avatar.displayName = 'Avatar'
