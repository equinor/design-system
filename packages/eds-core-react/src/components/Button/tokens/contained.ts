import { tokens } from '@equinor/eds-tokens'
import { ButtonToken } from '../Button.types'
import { button } from './button'

import { mergeDeepRight } from 'ramda'

const {
  colors: {
    text: {
      static_icons__primary_white: { rgba: primaryWhite },
    },
    interactive: {
      primary__resting: { rgba: primaryColor },
      primary__hover: { rgba: primaryHoverColor },
      secondary__resting: { rgba: secondaryColor },
      secondary__link_hover: { rgba: secondaryHoverColor },
      danger__resting: { rgba: dangerColor },
      danger__hover: { rgba: dangerHoverColor },
    },
  },

  shape: {
    button: { borderRadius: buttonBorderRadius },
  },
} = tokens

export const primary = mergeDeepRight(button, {
  background: primaryColor,
  typography: {
    color: primaryWhite,
  },
  border: {
    type: 'border',
    style: 'solid',
    width: '1px',
    color: primaryColor,
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
  },
  states: {
    hover: {
      background: primaryHoverColor,
    },
  },
}) as ButtonToken
export const secondary = mergeDeepRight(primary, {
  background: secondaryColor,
  border: {
    color: secondaryColor,
  },
  states: {
    hover: {
      background: secondaryHoverColor,
      border: {
        color: secondaryHoverColor,
      },
    },
  },
}) as ButtonToken

export const danger = mergeDeepRight(primary, {
  background: dangerColor,
  border: {
    color: dangerColor,
  },
  states: {
    hover: {
      background: dangerHoverColor,
      border: {
        color: dangerHoverColor,
      },
    },
  },
}) as ButtonToken
