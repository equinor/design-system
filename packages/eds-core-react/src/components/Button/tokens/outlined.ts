import { tokens } from '@equinor/eds-tokens'
import { mergeDeepRight } from 'ramda'
import { button } from './button'
import { ButtonToken } from '../Button.types'

const {
  colors: {
    interactive: {
      primary__resting: { rgba: primaryColor },
      primary__hover: { rgba: primaryHoverColor },
      primary__hover_alt: { rgba: primaryHoverAltColor },
      secondary__resting: { rgba: secondaryColor },
      secondary__link_hover: { rgba: secondaryHoverColor },
      secondary__highlight: { rgba: secondaryHoverAltColor },
      danger__resting: { rgba: dangerColor },
      danger__hover: { rgba: dangerHoverColor },
      danger__highlight: { rgba: dangerHoverAltColor },
    },
  },

  shape: {
    button: { borderRadius: buttonBorderRadius },
  },
} = tokens

export const primary = mergeDeepRight(button, {
  typography: {
    color: primaryColor,
  },
  border: {
    type: 'border',
    style: 'solid',
    width: '1px',
    color: primaryColor,
    radius: `var(--eds_button__radius, ${buttonBorderRadius})`,
  },
  states: {
    hover: {
      typography: {
        color: primaryHoverColor,
      },
      background: primaryHoverAltColor,
      border: {
        style: 'solid',
        type: 'border',
        width: '1px',
        color: primaryHoverColor,
      },
    },
    disabled: {
      background: 'transparent',
    },
  },
}) as ButtonToken

export const secondary = mergeDeepRight(primary, {
  typography: {
    color: secondaryColor,
  },
  border: {
    color: secondaryColor,
  },
  states: {
    hover: {
      background: secondaryHoverAltColor,
      typography: {
        color: secondaryHoverColor,
      },
      border: {
        color: secondaryHoverColor,
      },
    },
  },
}) as ButtonToken

export const danger = mergeDeepRight(primary, {
  typography: {
    color: dangerColor,
  },
  border: {
    color: dangerColor,
  },
  states: {
    hover: {
      background: dangerHoverAltColor,
      typography: {
        color: dangerHoverColor,
      },
      border: {
        color: dangerHoverColor,
      },
    },
  },
}) as ButtonToken
