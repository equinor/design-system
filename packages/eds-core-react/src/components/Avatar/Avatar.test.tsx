/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react'
import { axe } from 'jest-axe'
import styled from 'styled-components'
import { Avatar } from '.'
import { avatar as tokens } from './Avatar.tokens'

const StyledAvatar = styled(Avatar)`
  position: relative;
`

describe('Avatar', () => {
  it('Matches snapshot', () => {
    const src = 'https://i.imgur.com/UM3mrju.jpg'
    const { asFragment } = render(<Avatar alt="avatar" src={src} />)

    expect(asFragment()).toMatchSnapshot()
  })
  it('Should pass a11y test', async () => {
    const { container } = render(<Avatar alt="avatar" />)
    expect(await axe(container)).toHaveNoViolations()
  })
  it('Can extend the css for the component', () => {
    render(<StyledAvatar alt="avatar" data-testid="avatar" />)
    const avatar = screen.getByTestId('avatar')

    expect(avatar).toHaveStyleRule('position', 'relative')
  })
  it('Image has provided src', () => {
    const src = 'https://i.imgur.com/UM3mrju.jpg'
    render(<Avatar alt="avatar" src={src} />)
    const avatarImg = screen.getByAltText('avatar')

    expect(avatarImg).toHaveAttribute('src', src)
  })
  it('Has correct size', () => {
    const src = 'https://i.imgur.com/UM3mrju.jpg'
    const size = 32
    render(<Avatar src={src} size={size} alt="avatar" data-testid="avatar" />)
    const avatar = screen.getByTestId('avatar')

    expect(avatar).toHaveStyleRule('width', `${size}px`)
    expect(avatar).toHaveStyleRule('height', `${size}px`)
  })
  it('Has faded image when disabled', () => {
    const altText = 'avatar'
    const src = 'https://i.imgur.com/UM3mrju.jpg'
    render(<Avatar disabled src={src} alt={altText} />)
    const avatarImg = screen.getByAltText(altText)

    expect(avatarImg).toHaveStyleRule('opacity', tokens.states.disabled.opacity)
  })
})
