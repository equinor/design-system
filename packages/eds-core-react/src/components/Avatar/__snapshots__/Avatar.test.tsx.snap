// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  overflow: hidden;
  border-radius: 50%;
  height: 24px;
  width: 24px;
}

.c1 {
  height: 100%;
  text-align: center;
  color: transparent;
}

<div
    class="c0"
    size="24"
  >
    <img
      alt="avatar"
      class="c1"
      src="https://i.imgur.com/UM3mrju.jpg"
    />
  </div>
</DocumentFragment>
`;
